src/Common/tests/System/Collections/IDictionary.Generic.Tests.cs

```diff
@@ -0,0 +1,834 @@
+                dictionary[default(TKey)] = value;
+                Assert.Equal(value, dictionary[default(TKey)]);
+            }
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IDictionary_Generic_ItemSet_OnReadOnlyDictionary_ThrowsNotSupportedException(int count)
+        {
+            if (IsReadOnly)
+            {
+                IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+                TKey missingKey = GetNewKey(dictionary);
+                Assert.Throws<NotSupportedException>(() => dictionary[missingKey] = TValueFactory(5312));
+            }
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IDictionary_Generic_ItemSet_AddsNewValueWhenNotPresent(int count)
+        {
+            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+            TKey missingKey = GetNewKey(dictionary);
+            dictionary[missingKey] = TValueFactory(543);
+            Assert.Equal(count + 1, dictionary.Count);
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IDictionary_Generic_ItemSet_ReplacesExistingValueWhenPresent(int count)
+        {
+            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+            TKey existingKey = GetNewKey(dictionary);
+            dictionary.Add(existingKey, TValueFactory(5342));
+            TValue newValue = TValueFactory(1234);
+            dictionary[existingKey] = newValue;
+            Assert.Equal(count + 1, dictionary.Count);
+            Assert.Equal(newValue, dictionary[existingKey]);
+        }
+
+        #endregion
+
+        #region Keys
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IDictionary_Generic_Keys_ContainsAllCorrectKeys(int count)
+        {
+            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+            IEnumerable<TKey> expected = dictionary.Select((pair) => pair.Key);
+            Assert.True(expected.SequenceEqual(dictionary.Keys));
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IDictionary_Generic_Keys_ModifyingTheDictionaryUpdatesTheCollection(int count)
+        {
+            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+            ICollection<TKey> keys = dictionary.Keys;
+            dictionary.Clear();
+            Assert.Empty(keys);
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public virtual void IDictionary_Generic_Keys_Enumeration_ParentDictionaryModifiedInvalidates(int count)
+        {
+            IDictionary<TKey, TValue> dictionary = GenericIDictionaryFactory(count);
+            ICollection<TKey> keys = dictionary.Keys;
+            IEnumerator<TKey> keysEnum = keys.GetEnumerator();
+            dictionary.Clear();
+            // Undefined behavior
```

Comment:
Undefined, meaning it's not specified in the docs but our implementations are consistent?  Or meaning that our implementations vary wildly?  Ideally our tests would lock in a specific behavior, even if it's unspecified, so that we can easily find breaking changes in the future.